Declarative RLP Encoding/Decoding

emlautarom1 opened the pull request:
## Changes

- Introduce an alternative approach to RLP encoding and decoding, based on a declarative API with support for code generation through [Source Generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/)

## Types of changes

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

The core library has 100% test coverage. Source generated code might not be fully covered.

## Documentation

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

## Remarks

When we started working on refactoring our `TxDecoder` one thing that came up was how unergonomic is to work with our current RLP API. We even have some comments on the code itself mentioning these difficulties, for example:

https://github.com/NethermindEth/nethermind/blob/b81070de7a905b52a6f605ebdd4f417fefe234c7/src/Nethermind/Nethermind.Serialization.Rlp/Eip2930/AccessListDecoder.cs#L17-L23

https://github.com/NethermindEth/nethermind/blob/b81070de7a905b52a6f605ebdd4f417fefe234c7/src/Nethermind/Nethermind.Serialization.Rlp/Eip2930/AccessListDecoder.cs#L76-L81

This PR introduces a new RLP API based on https://github.com/NethermindEth/nethermind/pull/7334#issuecomment-2329147512 with several improvements:

- Describe the structure of a record and get encoding and decoding for free. No code duplication required.
- Records can be described using other records. Supports conditional, exceptions, function calls, etc.
- Decoding and encoding are extensible through classes that can be defined anywhere, plus some extension methods.
- Minimal core library with 100% code coverage.
- Supports backtracking.
- All function calls are known ahead of time (no `virtual` or `override`). Interfaces are only used to enforce implementations.
- Despite the extensive usage of lambdas, no closures are required (all lambdas are `static`). You can still use them if you want to, but overloads are provided to avoid them.
- Automatically generate the required code through Source Generators.

emlautarom1 committed:
Initial Rlp + Writer

emlautarom1 committed:
Rename `Sequence` -> `List`

emlautarom1 committed:
Initial `RlpReader`

emlautarom1 committed:
Initial `ReadList`

emlautarom1 committed:
Multiple `ReadList`

- Test for Set Theoretical Representation

emlautarom1 committed:
Restructure into separate files

- Extend instances

emlautarom1 committed:
Use `Rlp.Read` API

emlautarom1 committed:
Add `HeterogeneousList` test

emlautarom1 committed:
Add overload for `Action`

emlautarom1 committed:
Test `UnknownLengthList`

emlautarom1 committed:
Rename converter

emlautarom1 committed:
Use custom Exception

emlautarom1 committed:
Add `Action` overload

emlautarom1 committed:
Test for invalid readings

emlautarom1 committed:
Support long lists (+55 bytes)

emlautarom1 committed:
Implement interface on `ReadOnlySpanConverter`

emlautarom1 committed:
Make `Rlp[Reader|Writer]` symmetric

- Good things happen to those who respect symmetry.

emlautarom1 committed:
Support ref struct

emlautarom1 committed:
Reorder tests

emlautarom1 committed:
Annotate as scoped

- Code works, but compiler complains

emlautarom1 committed:
Initial `Choice` implementation

emlautarom1 committed:
Cleanup test

emlautarom1 committed:
Restructure internals

- Primitives are integers, byte sequences, and lists
- No need for `IntRlpConverter` (covered by primitives)

emlautarom1 committed:
Test for deep `Choice` (backtracking)

emlautarom1 committed:
Move `IRlpConverter`

emlautarom1 committed:
Demo user-defined record support

emlautarom1 committed:
Remove versions

emlautarom1 committed:
Use `ref struct` over `Interface`

- Replace virtual calls with conditional and inner state flag
- Refactor call sites

emlautarom1 committed:
Consistent error

emlautarom1 committed:
Split tests from library

- Rename to `FastRlp` to avoid conflicts

emlautarom1 committed:
Initial source generator for `RlpSerializable`

emlautarom1 committed:
Implement `Write`

emlautarom1 committed:
Test for Source Generated instances

emlautarom1 committed:
Move record

emlautarom1 committed:
No need to copy `using` directives

emlautarom1 committed:
Remove `Derived` namespace

- Derived code lives in the same namespace as the input record

emlautarom1 committed:
Update docs

emlautarom1 committed:
Rename `List` to `Sequence`

- Avoids conflicts with C#'s `List`

emlautarom1 committed:
Add support for `List` collection

emlautarom1 committed:
Test for equivalence in collections with explicit

emlautarom1 committed:
Add explicit converters for IntXX

- Required as Generics in Collections

emlautarom1 committed:
Reduce duplication

emlautarom1 committed:
Add `Dictionary` collection

emlautarom1 committed:
Extend generator to support Generics (WIP)

emlautarom1 committed:
Support multi-param generics

emlautarom1 committed:
Add Generic writer Action

- Required for Collections

emlautarom1 committed:
Rewrite Generics

- Remove Converter type parameters
- Blocks nested Collections

emlautarom1 committed:
Add recursive record test

emlautarom1 committed:
Enable `nullable`

emlautarom1 committed:
Fix `ReadBytes`

emlautarom1 committed:
Test for all base Integer types

emlautarom1 committed:
Initial README

emlautarom1 committed:
Make `Read` calls static

emlautarom1 committed:
Add `context` overload

- Allows for pure static lambdas

emlautarom1 committed:
Remove allocations

emlautarom1 committed:
Use `static` on `Write`

emlautarom1 committed:
Use `static` on `Read`

emlautarom1 committed:
Return `byte[]` instead of `ReadOnlySpan<byte>` in overload

emlautarom1 committed:
Use `static` on `Write`

emlautarom1 committed:
Fix typo

emlautarom1 committed:
Add overloads for `TContext`

emlautarom1 committed:
Use `context` overload when possible

emlautarom1 committed:
Remove duplicated overloads

emlautarom1 committed:
Remove "context" TODO

emlautarom1 committed:
Remove allocating overload

- Not actually needed in practice

emlautarom1 committed:
More `static` usage

emlautarom1 committed:
Use "context" overload in Generator

emlautarom1 committed:
Annotate as `scoped`

emlautarom1 committed:
Optimize `StringRlpConverter`

emlautarom1 committed:
Rename `FastRlp` to `FluentRlp`

- It's not that "fast"

emlautarom1 committed:
Remove allocations in Dictionary

emlautarom1 committed:
Remove unused method

emlautarom1 committed:
Introduce "context" overload for `Reader`

emlautarom1 committed:
Remove allocations in `Dictionary`

emlautarom1 committed:
Remove allocations in `List`

emlautarom1 committed:
Use field Encoding

emlautarom1 committed:
Add `ArrayRlpConverter`

emlautarom1 committed:
Remove allocations

emlautarom1 committed:
Add `TupleRlpConverter`

- Only 2-element tuples for now

emlautarom1 committed:
Support `Tuple` in source generators

emlautarom1 committed:
Add support for `Representation`

emlautarom1 committed:
Formatting

emlautarom1 committed:
Update docs

LukaszRozmej commented:
Can you move some RLP decoders to this new approach for testing purposes?
Can you share their generated code for inspection?

LukaszRozmej reviewed (commented):
Would be great to move from string builders to some more readable format. Even giant string interpolation would be good here.

emlautarom1 commented:
This is how the generated code looks like for the Rose Tree (formatted):

https://github.com/NethermindEth/nethermind/blob/8341a0e956b0c7995a2e50f3c1b8e4f9a5bf4310/src/Nethermind/Nethermind.Serialization.FluentRlp.Test/RlpDerivedTest.cs#L22-L23

```C#
// <auto-generated />
#nullable enable
using System;
using System.CodeDom.Compiler;
using Nethermind.Serialization.FluentRlp;
using Nethermind.Serialization.FluentRlp.Instances;

namespace Nethermind.Serialization.FluentRlp.Test;

[GeneratedCode("RlpSourceGenerator", "0.1")]
public abstract class TreeRlpConverter : IRlpConverter<Nethermind.Serialization.FluentRlp.Test.Tree>
{
    public static void Write(ref RlpWriter w, Nethermind.Serialization.FluentRlp.Test.Tree value)
    {
        w.WriteSequence(value, static (ref RlpWriter w, Nethermind.Serialization.FluentRlp.Test.Tree value) =>
        {
            w.Write(value.Value);
            w.Write(value.Children, static (ref RlpWriter w, Tree value) => { w.Write(value); });
        });
    }

    public static Nethermind.Serialization.FluentRlp.Test.Tree Read(ref RlpReader r)
    {
        return r.ReadSequence(static (scoped ref RlpReader r) =>
        {
            var Value = r.ReadString();
            var Children = r.ReadList(static (scoped ref RlpReader r) => { return r.ReadTree(); });
            return new Nethermind.Serialization.FluentRlp.Test.Tree(Value, Children);
        });
    }
}

[GeneratedCode("RlpSourceGenerator", "0.1")]
public static class TreeExt
{
    public static Nethermind.Serialization.FluentRlp.Test.Tree ReadTree(this ref RlpReader reader) => TreeRlpConverter.Read(ref reader);
    public static void Write(this ref RlpWriter writer, Nethermind.Serialization.FluentRlp.Test.Tree value) => TreeRlpConverter.Write(ref writer, value);
}
```

In Rider you can inspect the generated in the Solution explorer:

![image](https://github.com/user-attachments/assets/86abeecd-8d39-48d5-a544-1db2870ff0a8)

emlautarom1 committed:
Throw on trailing bytes

emlautarom1 commented on src/Nethermind/Nethermind.Serialization.FluentRlp/Rlp.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        var reader = new RlpReader(source);
+        T result = func(ref reader);
+        // TODO: We might want to add an option to check for no trailing bytes.
```
Added: now the `RLP` API will throw when there are trailing bytes.

emlautarom1 committed:
Reduce usage of `StringBuilder`

emlautarom1 committed:
Remove unused code

emlautarom1 committed:
Use format string

emlautarom1 committed:
Merge `Array` with generics

emlautarom1 commented:
Updated generators to use string interpolation. There are some places where we still use `StringBuilder` though.

emlautarom1 committed:
Formatting

LukaszRozmej commented:
> This is how the generated code looks like for the Rose Tree (formatted):

Ok but can we do something real-life for us? Like `BlockHeader` for example?

LukaszRozmej commented on src/Nethermind/Nethermind.Serialization.FluentRlp/RlpWriter.cs:
```diff
@@ -0,0 +1,203 @@
+
+    public int Length { get; private set; }
+
+    private byte[] _buffer;
```
We often write into Netty arena-based buffers to avoid allocations, would be good to support that.

LukaszRozmej reviewed (commented):
What about `RlpBehaviors`?

Scooletz commented on src/Nethermind/Nethermind.Serialization.FluentRlp/Int32Primitive.cs:
```diff
@@ -0,0 +1,23 @@
+    public static int Read(ReadOnlySpan<byte> source)
+    {
+        Span<byte> buffer = stackalloc byte[sizeof(Int32)];
+        source.CopyTo(buffer[^source.Length..]);
```
`Unsafe.ReadUnalaligned`? Why copy?

emlautarom1 replied:
This method was added due to `BinaryPrimitives.ReadInt32BigEndian` requiring exactly 4 byte to read a `Int32`, so we pad `source` with enough `0` so it can properly decode.

I've never used `Unsafe.ReadUnaligned`, not even `usafe` code. How would that look?

LukaszRozmej replied:
Why not use `Bytes` class we already have?

emlautarom1 replied:
Didn't think about it, but seems like implementations are the same:

https://github.com/NethermindEth/nethermind/blob/c8748770fe88fecf3b0b36226296e5c0858da051/src/Nethermind/Nethermind.Core/Extensions/Bytes.cs#L439-L441

Seems a bit overkill to add a project reference for 3 LOCs.

Scooletz commented on src/Nethermind/Nethermind.Serialization.FluentRlp/Rlp.cs:
```diff
@@ -0,0 +1,33 @@
+        action(ref lengthWriter, ctx);
+        var serialized = new byte[lengthWriter.Length];
+        var contentWriter = RlpWriter.ContentWriter(serialized);
+        action(ref contentWriter, ctx);
```
When is it used? Do we write the data twice?

emlautarom1 replied:
When is what used? We "write" the data twice to first compute the length (`LengthWriter`), and then we actually write the bytes into a buffer (`ContentWriter`).

Scooletz replied:
This is costly. There might be a way to write it once if you do it in a way similar to Paprika. In Paprika, there's enough room left at the beginning to encode any length, then once the encoding is done, the length is written at the beginning and the position is adjusted. This would require a copy of the memory though as it would leave a gap at the beginning. This might be still cheaper if you copy the memory once more instead of going through the serialization two times.

Having this said, maybe adding some context instead of a writer could be useful. If there was a context you could borrow from it an array to write to, without getting into ArrayPool.Shared thing. In Paprika profiling this was costly and was eliminated by passing a context object that has a simple stack of arrays that you can rent from (single threaded, no need to deal with the shared thing).

https://github.com/NethermindEth/Paprika/blob/61c326a39e942ab9c550f6a6b9b5df59e67c0f0f/src/Paprika/Merkle/ComputeMerkleBehavior.cs#L750-L757

emlautarom1 replied:
This could work but I think it would be quite a redesign if it were to be adapted to all paths, not only here at the top level. Recursive calls in `WriteSequence` would need to be adjusted to first immediately write the bytes and then move them around to deal with writing the length first.

Could be worth exploring though.

Scooletz replied:
Moving once vs writing twice 👀 Clearly, I'm just asking as I don't have a clear answer. Also, in Paprika there are just a few cases where it's needed as majority of the written RLP is really short (beside branches, where it's applied).

Scooletz commented on src/Nethermind/Nethermind.Serialization.FluentRlp/RlpWriter.cs:
```diff
@@ -0,0 +1,203 @@
+        };
+    }
+
+    public static RlpWriter ContentWriter(byte[] buffer)
```
Have you considered adding an option aligned with `Utf8JsonWriter` where there's a ctor that accepts `IBufferWriter<byte>`? This would mean that ufortunately dicontinued chunks should be supported, but could allow to provide a writer over anything. Maybe this could help to address Netty comment from @LukaszRozmej  

https://learn.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonwriter.-ctor?view=net-9.0#system-text-json-utf8jsonwriter-ctor(system-buffers-ibufferwriter((system-byte))-system-text-json-jsonwriteroptions)

emlautarom1 replied:
Interesting. I picked `byte[]` as a safe default but there is no reason why other type could be used.

- `IBufferWriter<T>` is quite small, it's part of the std lib and it supports `Span`-based APIs.

- `NettyRlpStream` is based on `IByteBuffer`, the latter which is defined in `DotNetty.Buffers`. It supports the operations that we need but it's quite large.

Now, the issue is that `IByteBuffer` and `IBufferWriter` are unrelated so we would need to pick one (unless we start doing conversions).

LukaszRozmej replied:
`IBufferWriter` is probably more useful, we can try making adapter to `IByteBuffer` if we want

emlautarom1 replied:
Replaced `byte[]` buffer with `IBufferWriter<byte>`.

The return type of the `Rlp` interface is now `ReadOnlyMemory<byte>` instead of `byte[]` for more flexibility compared to returning `ReadOnlySpan<byte>`. Overloads for `Rlp.Read(ReadOnlyMemory<byte>)` were added.

Scooletz reviewed (commented):
What an interesting take on this topic 😍 A few remarks provided as comments. General:

1. Probably `[SkipLocalInit]` would be beneficial.
2. Benchmarking with potential ASM output
3. Converting some of the existing ones and performing mano-a-mano comparison.

Scooletz commented on src/Nethermind/Nethermind.Serialization.FluentRlp.Generator/RlpSourceGenerator.cs:
```diff
@@ -0,0 +1,320 @@
+
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var provider = context.SyntaxProvider.CreateSyntaxProvider(
```
Consider attribute based creation using [ForAttributeWithMetadataName](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.syntaxvalueprovider.forattributewithmetadataname?view=roslyn-dotnet-4.9.0). It should be much more cheaper than scanning all the records and only then select on attribute basis.

emlautarom1 committed:
Fix `0`

emlautarom1 committed:
Remove signed integer requirement

emlautarom1 committed:
Add `BytesRead`

emlautarom1 commented:
@LukaszRozmej `RlpBehaviors` as currently defined are not supported by the API though you can get the same behavior by manually pass any "context" during read/write.

We recently changed the `Rlp` interface so trailing bytes throw by default (though you can avoid it by using the `RlpReader`/`RlpWriter` structs directly if needed).

emlautarom1 commented:
@LukaszRozmej `RlpBehaviors` are not explicitly supported by the API but you can get the same behavior by manually passing any "context" when reading/writing.

Note that we recently changed the `Rlp` interface so trailing bytes throw by default. If you want more control over what happens before or after reading/writing you can use the `RlpReader` and `RlpWriter` APIs directly.

emlautarom1 committed:
Include user `using` statements

emlautarom1 committed:
Fix nested Generics support

emlautarom1 committed:
Initial benchmark

emlautarom1 commented:
I've added a benchmark that encodes and decodes an `AccessList` as defined in:

https://github.com/NethermindEth/nethermind/blob/e0c4a599d8f0ee1c8e3ce13d1c8620d4e3255a23/src/Nethermind/Nethermind.Core/Eip2930/AccessList.cs#L15

Results on my machine are the following:

```
| Method  | Mean     | Error   | StdDev  | Ratio |
|-------- |---------:|--------:|--------:|------:|
| Current | 343.9 us | 1.43 us | 1.34 us |  1.00 |
| Fluent  | 834.9 us | 2.34 us | 2.19 us |  2.43 |
```

There is room for a possible optimization: some records like `Address` have a known fixed byte size which we can leverage to avoid having to copy bytes twice: once to figure out the length and the other to actually copy the bytes.

emlautarom1 committed:
Prefer `sizeof(T)` over `Marshal.SizeOf<T>`

- Nice speedup

emlautarom1 commented:
Replacing `Marshal.SizeOf<T>()` with `sizeof(T)` and some unsafe annotations gives quite the boost at no cost:

```
| Method  | Mean     | Error   | StdDev  | Ratio | RatioSD |
|-------- |---------:|--------:|--------:|------:|--------:|
| Current | 359.8 us | 5.03 us | 4.70 us |  1.00 |    0.02 |
| Fluent  | 626.2 us | 2.90 us | 2.42 us |  1.74 |    0.02 |
```

emlautarom1 committed:
Formatting

emlautarom1 committed:
File encoding

emlautarom1 committed:
Remove unused import

emlautarom1 commented on src/Nethermind/Nethermind.Serialization.FluentRlp/RlpWriter.cs:
```diff
@@ -0,0 +1,203 @@
+        var size = sizeof(T);
+        Span<byte> bigEndian = stackalloc byte[size];
+        value.WriteBigEndian(bigEndian);
+        bigEndian = bigEndian.TrimStart((byte)0);
```
`TrimStart` does not seem to be heavily optimized. There might be something better that we can use, specially considering that we're removing leading zeros.

LukaszRozmej replied:
Looks like .net doesn't use SIMD for it!
We could write Vector based way to find start index.
@benaadams

emlautarom1 commented:
After running some benchmarks I found that `UInt256` was getting boxed due to the usage of interface default bodies. Fixing that issue improves performance and drastically reduces memory allocation (added `[MemoryDiagnoser]` as requested by @Scooletz):

```
| Method  | Mean     | Error   | StdDev  | Ratio | Gen0     | Gen1     | Gen2     | Allocated  | Alloc Ratio |
|-------- |---------:|--------:|--------:|------:|---------:|---------:|---------:|-----------:|------------:|
| Current | 359.5 us | 0.98 us | 0.86 us |  1.00 | 166.5039 | 166.5039 | 166.5039 | 1033.87 KB |        1.00 |
| Fluent  | 530.1 us | 4.67 us | 4.14 us |  1.47 |  51.7578 |  51.7578 |  51.7578 |   617.5 KB |        0.60 |
```

emlautarom1 committed:
Avoid boxing due to interface default bodies

emlautarom1 committed:
Add `MemoryDiagnoser`

Scooletz commented on src/Nethermind/Nethermind.Serialization.Rlp.Benchmark/Program.cs:
```diff
@@ -0,0 +1,120 @@
+        var decoder = Eip2930.AccessListDecoder.Instance;
+
+        var length = decoder.GetLength(_current, RlpBehaviors.None);
+        var stream = new RlpStream(length);
```
This line is responsible for all the allocations as it allocates a new array underneath https://github.com/NethermindEth/nethermind/blob/4e4455960a4d62ea3429e7bb143b45789ec7246f/src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs#L44 Is this the case that we want to benchmark or rather it should be a reused `RlpStream` here?

emlautarom1 replied:
To be comparable with the `FluentRlp` approach we should allocate a new buffer. Since both will allocate the same buffer size then it should not matter.

Scooletz replied:
I see. If they allocate the same buffer, what makes the current allocate over 400kb more then?

emlautarom1 replied:
Interestingly they're not allocating the same buffer size: the fluent approach uses a buffer of `170850` bytes while the current one uses `172845`. That does not account for the 400kb you mention though. Rider's profiler is not giving me anything useful so I'm kind of stuck now.

Maybe we should add other objects (ex. `LogEntry`, `BlockInfo`, etc.) to get more accurate benchmarks.

LukaszRozmej replied:
you can use NettyRlpStream that uses arena memory

emlautarom1 committed:
Use `IBufferWriter` instead of `byte[]`

- Return value is now `ReadOnlyMemory<byte>`
- Add overloads for reading `ReadOnlyMemory<byte>`
- Add `FluentAssertions` extensions

emlautarom1 committed:
Use `IBufferWriter` instead of `byte[]`

LukaszRozmej commented on src/Nethermind/Nethermind.Serialization.FluentRlp/Rlp.cs:
```diff
@@ -37,3 +33,43 @@ public static T Read<T>(ReadOnlySpan<byte> source, RefRlpReaderFunc<T> func)
+    /// <param name="capacity">The capacity of the underlying buffer.</param>
+    public FixedArrayBufferWriter(int capacity)
+    {
+        _buffer = new T[capacity];
```
ArrayPool?

emlautarom1 replied:
By default I think we should go with a plain array to match the `ArrayBufferWriter` behavior. At the end of the day, the `RLP` static class is like a "safe default" API.

If we want more control over the buffers we use we can write a custom `IBufferWriter` as you suggested earlier while using `RlpReader` and `RlpWriter` directly.

emlautarom1 committed:
Add support for `Optional`

emlautarom1 committed:
Use `sizeof` instead of `Marshal.SizeOf<T>` in `RlpReader`

emlautarom1 committed:
More tests for `Optional`

emlautarom1 committed:
Include projects in slnx

emlautarom1 committed:
Adjust csproj files

- Disable implicits
- Adjust target framework for source generator
- Use common testing properties

emlautarom1 committed:
Initial `length` option

- Improves code generation when a known length is provided

emlautarom1 committed:
Make test match benchmark

emlautarom1 committed:
Add length hint to source generator

emlautarom1 committed:
Formatting

emlautarom1 committed:
Reduce code duplication

emlautarom1 committed:
Expose internals as `Unsafe*`

emlautarom1 committed:
Fix debug asserts

emlautarom1-agent[bot] committed:
Conform FluentRlp to repo analyzer rules

Spell out types instead of `var` (IDE0008), use expression bodies (IDE0022)
and a primary constructor (IDE0290), all of which the repo enforces as build
errors.

emlautarom1-agent[bot] committed:
Assert with NUnit constraints in FluentRlp tests

The repo carries no FluentAssertions dependency. Structural comparisons of
records holding collections use `UsingPropertiesComparer()`, since record
equality compares collection members by reference.

emlautarom1-agent[bot] committed:
Benchmark FluentRlp against the current RLP decoder API

`Nethermind.Serialization.Rlp` exposes its own `RlpReader`/`RlpWriter`, so a
namespace nested under it shadows the FluentRlp types of the same name. Name
the project and namespace after what it benchmarks so both stay reachable.

Encode through `RlpDecoder<T>.Encode<TWriter>` and decode from the encoded
span, matching how callers drive the decoder.

emlautarom1-agent[bot] committed:
Encode 55-byte payloads with the short RLP prefix

RLP introduces payloads of 0 to 55 bytes inclusive with a single-byte prefix
and only reaches for the long form at 56. Sending 55 down the long form emits
non-canonical output that `Nethermind.Serialization.Rlp` rejects, since it
requires a long-form content length of at least `SmallPrefixBarrier` (56).

Applies to byte strings and sequences on both the length and content paths.

emlautarom1-agent[bot] committed:
Reject trailing bytes inside RLP sequences

A sequence reader is scoped to that sequence's payload, so one that stops before
the end has left items undecoded. Treating that as a successful decode accepts
malformed input, which is reachable from peer-supplied data.

`Rlp.Read` enforces this for the outermost value; extend it to every nested
sequence, on both the short and long prefix form.
